compiler/noirc_evaluator/src/ssa.rs

```diff
@@ -91,17 +91,136 @@ pub struct SsaEvaluatorOptions {
+        SsaPass::new(Ssa::make_constrain_not_equal_instructions, "Adding constrain not equal"),
+        SsaPass::new(Ssa::check_u128_mul_overflow, "Check u128 mul overflow"),
+        SsaPass::new(Ssa::dead_instruction_elimination, "Dead Instruction Elimination"),
+        SsaPass::new(Ssa::simplify_cfg, "Simplifying"),
+        SsaPass::new(Ssa::array_set_optimization, "Array Set Optimizations"),
+        // The Brillig globals pass expected that we have the used globals map set for each function.
+        // The used globals map is determined during DIE, so we should duplicate entry points before a DIE pass run.
+        SsaPass::new(Ssa::brillig_entry_point_analysis, "Brillig Entry Point Analysis"),
+        // Remove any potentially unnecessary duplication from the Brillig entry point analysis.
+        SsaPass::new(Ssa::remove_unreachable_functions, "Removing Unreachable Functions"),
+        SsaPass::new(Ssa::remove_truncate_after_range_check, "Removing Truncate after RangeCheck"),
+        // This pass makes transformations specific to Brillig generation.
+        // It must be the last pass to either alter or add new instructions before Brillig generation,
+        // as other semantics in the compiler can potentially break (e.g. inserting instructions).
+        // We can safely place the pass before DIE as that pass only removes instructions.
+        // We also need DIE's tracking of used globals in case the array get transformations
+        // end up using an existing constant from the globals space.
+        SsaPass::new(Ssa::brillig_array_gets, "Brillig Array Get Optimizations"),
+        SsaPass::new(Ssa::dead_instruction_elimination, "Dead Instruction Elimination"),
+    ]
+}
+
+/// The second SSA pipeline which executes Brillig functions from the primary
+/// pipeline to fold constants.
+pub fn secondary_passes(brillig: &Brillig) -> Vec<SsaPass> {
+    vec![
+        SsaPass::new(
+            move |ssa| ssa.fold_constants_with_brillig(brillig),
+            "Inlining Brillig Calls Inlining",
+        ),
+        // It could happen that we inlined all calls to a given brillig function.
+        // In that case it's unused so we can remove it. This is what we check next.
+        SsaPass::new(Ssa::remove_unreachable_functions, "Removing Unreachable Functions"),
+        SsaPass::new(Ssa::dead_instruction_elimination_acir, "Dead Instruction Elimination"),
+    ]
+}
+
 /// Optimize the given program by converting it into SSA
 /// form and performing optimizations there. When finished,
 /// convert the final SSA into an ACIR program and return it.
 /// An ACIR program is made up of both ACIR functions
 /// and Brillig functions for unconstrained execution.
-pub(crate) fn optimize_into_acir(
+pub(crate) fn optimize_into_acir<S>(
     program: Program,
     options: &SsaEvaluatorOptions,
-) -> Result<ArtifactsAndWarnings, RuntimeError> {
+    primary: &[SsaPass],
+    secondary: S,
```

Comment:
Please document the difference between `primary` and `secondary` here

Reply:
Done PTAL